samza-yarn/src/main/java/org/apache/samza/job/yarn/YarnClusterResourceManager.java

```diff
@@ -331,9 +332,24 @@ public void launchStreamProcessor(SamzaResource resource, CommandBuilder builder
   public void stopStreamProcessor(SamzaResource resource) {
     synchronized (lock) {
       Container container = allocatedResources.get(resource);
+      /*
+       * 1. Stop the container through NMClient if the container was instantiated as part of NMClient lifecycle.
+       * 2. Stop the container through AMClient by release the assigned container if the container was from the previous
+       *    attempt and managed by the AM due to AM-HA
+       * 3. Ignore the request if the container associated with the resource isn't present in the bookeeping.
+       */
       if (container != null) {
         log.info("Stopping Container ID: {} on host: {}", resource.getContainerId(), resource.getHost());
         this.nmClientAsync.stopContainerAsync(container.getId(), container.getNodeId());
+      } else {
+        YarnContainer yarnContainer = state.runningProcessors.get(resource.getContainerId());
+        if (yarnContainer != null) {
+          log.info("Stopping container from previous attempt with Container ID: {} on host: {}",
+              resource.getContainerId(), resource.getHost());
+          amClient.releaseAssignedContainer(yarnContainer.id());
```

Comment:
after amClient.releaseAssignedContainer(yarnContainer.id()); is done, what do we get the call back as?
because we should use that call back to restart the container right (as earlier we allocated resources only for processors not runnning in prev attempt).

Reply:
It follows the similar workflow as `StopStreamProcessor` which is notified through the AMRM client callback `onContainersCompleted` which will treat it as the container was stopped agnostic to which workflow triggered it and hence you get all the niceties of state update and instantiating new container in its place.

Reply:
awesome! thanks for telling me.